fix non-strict placeholder naming with kwargs

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144278

Fixes https://github.com/pytorch/pytorch/issues/143732

Differential Revision: [D67872055](https://our.internmc.facebook.com/intern/diff/D67872055/)

avikchaudhuri committed:
fix non-strict placeholder naming with kwargs

Fixes https://github.com/pytorch/pytorch/issues/143732

Differential Revision: [D67872055](https://our.internmc.facebook.com/intern/diff/D67872055/)

[ghstack-poisoned]

yushangdi commented on torch/_export/utils.py:
```diff
@@ -821,6 +821,14 @@ def _extract_pytree_key(x):
     combined_args = _bind_signature_to_inputs(mod, fake_args, fake_kwargs)
+    # reorder following args, kwargs
+    combined_args = {
+        name: combined_args[name]
```
just curious why the order in the dictionary matters? If the order matters, maybe we should make it an ordered dict to be explicit?

Skylion007 replied:
Use itertools.chain. No need for this weird list unpacking

avikchaudhuri replied:
Turns out that the graph signature has user inputs in the order they were received, but combined args orders them in the order of the original forward signature. The actual ordering doesn't matter so much as that they should be consistent, because we zip them soon afterwards.

avikchaudhuri replied:
@yushangdi I ended up changing the implementation a bit because it was not accounting for variadic kwargs. Latest update matches what happens in strict.

Given a module signature: `def forward(a, b, c, kwargs): ...`
and args: `a_val`, kwargs: `{c=c_val, b=b_val, d=d_val}`

we should now get placeholder names `a, c, b, d`. So overall preserving call site order instead of def site order.

avikchaudhuri committed:
Update on "fix non-strict placeholder naming with kwargs"


Fixes https://github.com/pytorch/pytorch/issues/143732

Differential Revision: [D67872055](https://our.internmc.facebook.com/intern/diff/D67872055/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "fix non-strict placeholder naming with kwargs"


Fixes https://github.com/pytorch/pytorch/issues/143732

Differential Revision: [D67872055](https://our.internmc.facebook.com/intern/diff/D67872055/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "fix non-strict placeholder naming with kwargs"


Fixes https://github.com/pytorch/pytorch/issues/143732

Differential Revision: [D67872055](https://our.internmc.facebook.com/intern/diff/D67872055/)

[ghstack-poisoned]

yushangdi reviewed (approved):
LGTM, thanks!

pianpwk reviewed (approved):
looks great!

avikchaudhuri commented:
@pytorchbot merge -f "bogus CI failures"
